schema-layer/data-structures/generics.md

```diff
@@ -0,0 +1,109 @@
+# Specification: IPLD Generics
+
+**Status: Prescriptive - Exploratory**
+
+Organizing IPLD data into usable, efficient, complex data structures spanning many blocks aimed for use by end-user applications.
+
+This document will re-use some terms found in the [IPLD data model](/IPLD-Data-Model-v1.md). 
+
+IPLD Generics offer codec agnostic programming interfaces for all common operations users can currently accomplish on Data-Model [“Kinds”]().
+
+Contents:
+
+  * [Motivation](#motivation)
+  * [Requirements](#requirements)
+  * [Operations](#operations)
+  * [Implementations](#implementations)
+
+## Motivation
+
+Even before the IPLD Data-Model was formally specified developers were creating multi-block data-structures with similar semantics to single-block primitives. The most illustrative example of this is the `dag-pb` HAMT implementation used by IPFS for large directories.
+
+These early implementations of multi-block data structures exposed several problems.
+
+  * They are lacking self-description. A consumer of a graph containing these structures would have to have logic on top of IPLD and vary the way it performs operations on there data structures.
+  * Implementations of these data structures cannot perform operations on each other. In other words, multi-block data structures have a hard time building on top of each other.
+
+Since there wasn’t a standardized way to describe these data structures we couldn’t build libraries for paths and selectors that seamlessly supported them.
+
+As we started designing this system several other requirements surfaced.
+
+* Transparent encryption envelopes on the read and write side.
+* Advanced `Link` types that can support some form of mutability and link to paths within other data structures.
+* Flexible multi-block binary types.
+
+## Requirements
+
+IPLD Generics cannot be implemented without:
+
+  * The IPLD Data-Model. There are codec agnostic but do require that the full data model be implemented by the codec.
+  * Reserved `_type` property.
```

Comment:
We talked about this irl the other day, but for internet posterity: I don't actually think this is a strict requirement; and, it has huge drawbacks.

There are at least three different ways I can think of to solve the signalling problem:

1. when writing code against the libraries, simply Do Things as intended; signaling isn't necessary at all.  (This is defacto what everything previously as done, and we'll never *stop* it from being doable; though we may want to also offer *better* alternatives that are more communicable rather than implementation specific.)

2. Use schemas as a signalling mechanism -- some description at http://slides.com/warpfork/ipld-gpn-2019#/9/24 as well as our other various gists and drafts in recent weeks.  This has many boons:
  - it's not in-band signalling (not blue-box prone; requires no reserved words and causes no asciifucking or escaping requirements)...!
  - it provides documentation of what's going on (remember, one of the goals of the schema system as a whole is to make it possible to describe systems well from a high level; such descriptions would be woefully incomplete if they *didn't* include some definition of any sharding going on, so we **will** have this information there no matter what)
  - **it makes it possible to configure the libraries to produce *new data* without having do that configuration in an library/implementation-specific way.**  This is super important: if you're writing new data, where's the "_type" keyword going to come from?  It doesn't exists; there's no in-band signalling data at all when you're writing new data, so we need *something* here or we're just not feature complete! (We could try to separate out the concept of declarative formats for stating what a sharding or encrypting etc feature is going to do from schemas, and then just embed that in schemas again, if that makes this easier to think about.)
  - it's this concept of indicating the advanced layout strategies (or, not!) from the outside that enables features like choosing whether to path through things transparently at the high level or as easily address lower level content blocks: http://slides.com/warpfork/ipld-gpn-2019#/9/25 -- and remember, we need this to be serializable so that we can attach it to things like selectors sent over the wire.  (It's at this point that schemas become relevant again even if in the last bullet point, we separated out the declarative sharing config from schemas: to define *where* we're going to use a declarative sharding config, we need some recursive definition language, right?)

3. in-band signalling.

It's only the third of these options for signalling which requires a reserved property at all; therefore, I don't know if we should say we require a reserved property.

There's also *so many* upsides to the option 2 of signalling that I seriously question whether we need or want option 3 at all.  Especially if option 3 means we would lose the ability to use certain strings as map keys, which is... ouch.

Reply:
(Okay, this will be getting on a tangent, but I was thinking about the chooseable-transparency thing some more and schemas alone might not *always* do the thing we want.  I've mostly thought about that in the context of UnixFSv2 where we'll sometimes want to see "through" files to the raw blocks so we can get e.g. streaming the start of the file.  And in that case, it's clear how sending a different schema attached to a selector would allow that; files are a leaf.  But what if we wanted to do something like path normally through two directories, and then ask for only some shards of a third?  Since dirs are recursively defined in the hypothetical unixfsv2 schema, this is trickier: we might need something else: say, perhaps sending a replacement schema (which would switch the defn of dirs to the raw blocks rather than the advanced layout) that should only kick in beyond a certain path?  But this is complicated and not a thing we're trying to definitively solve today, so i'm gonna share but then park this train of thought.)

Reply:
Another update from our in-person chat, while the reservation “/“ is a little problematic, it is *already* reserved for `dag-json` and is a problematic key for paths anyway. Rather than reserving yet-another-keyword I think it best to move this from `_type` to `{“/“: {_type: ... }}`.

Reply:
I’ll get into more details later, but I’m not at all convinced that we don’t need in-band signaling. I’ll write up the motivations in more detail (TLDR; you can’t effectively link to other people’s data freely if that data isn’t self-describing, you’d have to add application logic to layer on the out-of-band schemas for any data you ever link to), but for now I’m going to continue to detail this method of self-description.

I don’t think this is the **only** signaling method, but it’s one that we need to support. And, the other forms of signaling we’re considering with schemas are built quite nicely on top of it (you just use a schema to set the reserved property, you wouldn’t need any special syntax).

Reply:
> It's only the third of these options for signalling which requires a reserved property at all; therefore, I don't know if we should say we require a reserved property.

I should adjust the language here to make this more clear. By “reserved” I do not mean “this is the only way to signal the implementation of a generic.” What I actually mean is “do not use this property for anything *other* than this signaling” because if they do things will break when trying to read this data in APIs that support IPLD Generics.

Reply:
The problem here is that you're talking about creating a reserved key *at the data model layer*, and we can't do that and keep our shorts on.

>  already reserved for dag-json

1. First and foremost, we mostly regret that.  There's conversations about how to mitigate and hopefully move away from this; we don't want to start another one.
2. the operative wording is "in dag-json" -- It's a concern that's completely encapsulated **in *that instance of* a codec, at the codec layer** and ***not*** in the data model.  That makes it different from what you're proposing here, because it's got a significantly more limited splash radius and we actually *do* have ways to move away from it.
3. We mostly figure on marking dag-json as a debugging-only not-for-all-data extremely-caution-before-using format because of this asciifraking incompleteness.  We can do this -- effectively, flag dag-json as a not-quite-'native' (e.g. doesn't support all values) format -- and I think we should do so; it's honest.  Further, @Stebalien has seriously floated that we might correspondingly propose a new json-esque format that uses deeper nesting of objects simply to ensure it doesn't violate control-plane-vs-data-plane.  (I don't think getting someone to actually implement that is a high priority, but I do think it's a good idea and sets the goalposts in the right places.)

>  is a problematic key for paths anyway

We can *do* escaping in paths.  This is simple (if in the 'SMOP' sense): define paths as a series of segments which are opaque strings; define the path representation as a string itself to be slash-joined plus escaping.

I'm not super worried about this.

We've used the pattern of "here's the canonical 'AST' form; here's a string simplification that's usually usable and definitely well-defined" in several places already.  We can use it again here.

> you can’t effectively link to other people’s data freely if that data isn’t self-describing, you’d have to add application logic to layer on the out-of-band schemas for any data you ever link to

This is a real and valid concern for sure.  It's also one we can solve.  Reserved words at the data model layer which either asciifrak us or introduce escaping _at the data model layer, regardless of codecs_, is *not* something we can solve again once introducing it.

We're already doing this pattern of application logic handling it in e.g. how the gateways turn URL paths into unixfsv1 lookups.  If we can do this again, but with standard conventions and library features rather than one-offs of that particular application, then we're winning.

I think the way you phrased that actually hints at happier solution paths: if the data isn't self-describing, we can provide the description with the link.  This is as good a place as any to provide the description, isn't it?

I wonder if we can come up with some more ways to describe this without using the term "out-of-band"?  "Out-of-band" seems like a pretty sharp-edged thing, and istm we've got more of a gradient: since our links are 'in-band' themselves, if we're associating data with them, is that really out-of-band in any negative ways?

(Maybe a helpful analogy, maybe not: This draws up associations in my mind of when someone was asking how we might for example associate temperature units with numbers: what if "C" and "F" isn't expressive enough -- what if someone uses "C" to mean something *else* that's not even a temperature unit?  The answer to this *isn't* to try reach into the ether and try to conjure Gödel numbers unique to the universe they're embedded in; the answer is "context matters", we can use other data from earlier in the parse to feel more confident we know something about what we're looking at, and *it's okay* lean on that.  Things get worse when we try to deny context; things get more tractable when we acknowledge it and put it to work.)

---

tl;dr and with vigor: Escaping of keys in the data model is a no fly zone.  The compat story with other formats and existing data will be insane.  The performance will be insane (half the reason CBOR is fast is there's no escaping to check for wasting O(n) cycles on every string!).  Everything would become a fractal of bad.  It's pandora's box.  We cannot open it and still produce a viable standard.

Defining escaping for string-represented paths is viable in ways that escaping or reserving keys in the data model itself is not.  If there's a tug of war between those two: escape paths.

Reply:
Working on a spec refactor that will:

* Split “Generic Definition” from the signal.
* Add an “open problem” section on how to best accomplish “in-data signalling”